File: 005_rust_ducks/src/move_strategy.rs
use crate::duck_traits::Duck;

pub struct SwimStrategy;
pub struct FlyStrategy;

pub trait MoveStrategy {
    fn moves<T: Duck>(t: &T) -> &str;
}

impl MoveStrategy for SwimStrategy {
    fn moves<T: Duck>(_t: &T) -> &str {
        "swims"
    }
}
impl MoveStrategy for FlyStrategy {
    fn moves<T: Duck>(_t: &T) -> &str {
        "flies"
    }
}

File: 005_duck_factory/src/main/java/com/github/kgrech/ducks/strategy/MoveStrategy.java
package com.github.kgrech.ducks.strategy;

import com.github.kgrech.ducks.Duck;

public enum MoveStrategy {
    FlyStrategy {
        @Override
        public String move(Duck duck) {
            return "flies";
        }
    },
    SwimStrategy {
        @Override
        public String move(Duck duck) {
            return "swims";
        }
    };

    public abstract String move(Duck duck);
}
